### PR TITLE
Reduce MoE activation memory in DualPipeV

### PR DESCRIPTION
**Pull request by MasterJH5574**
- Remove EpilogOuts/epilog_b (dead code; logits not needed after loss)
- Remove Stage2/Stage4 args and outs (only ctx needed for a2a backward)
- Add padded_index_gather: avoids saving input for backward + reduces CUDA allocator fragmentation via fixed-alignment output buffers
- Free stage-boundary tensor storage early via untyped_storage().resize_(0) after async all-to-all consumers complete (sorted_tokens after Stage 2, gathered_tokens after Stage 3, stage3 moe_outs after Stage 4)
- Add fwd_comm_deferred_free to ExecutionCtx for clean deferred-free in the overlapped path without changing stage function signatures
- Add layer_partition() for memory-aware pipeline stage assignment
- Add per-layer saved-tensor profiling with weight/activation distinction

Co-authored-by: Hao Kang <89672451+haok1402@users.noreply.github.com>

### CODE REVIEW

File: pithtrain/dualpipe/layer_partition.py
Comment (haok1402) [inline pithtrain/dualpipe/layer_partition.py:1]: @MasterJH5574 do we still need to keep the layer partition if the donation isn't used?

**Comment by haok1402** [comment]
The cross entropy loss using 4k sequence looks correct after a couple steps from the released checkpoint.

```
2026-04-12 22:39:00 | INFO | step 00000011/00000015 | step-time 17.691 sec | cross-entropy-loss 2.5770 | learning-rate 1.000000e-06 | gradient-norm 696.0355 | tokens-per-second 237,090 | peak-gpu-memory 58.31 GB
2026-04-12 22:39:18 | INFO | step 00000012/00000015 | step-time 17.741 sec | cross-entropy-loss 2.5923 | learning-rate 1.000000e-06 | gradient-norm 1003.7461 | tokens-per-second 236,419 | peak-gpu-memory 58.37 GB
2026-04-12 22:39:36 | INFO | step 00000013/00000015 | step-time 18.392 sec | cross-entropy-loss 2.6027 | learning-rate 1.000000e-06 | gradient-norm 1149.4879 | tokens-per-second 228,045 | peak-gpu-memory 58.51 GB
2026-04-12 22:39:54 | INFO | step 00000014/00000015 | step-time 17.032 sec | cross-entropy-loss 2.5203 | learning-rate 1.000000e-06 | gradient-norm 713.4630 | tokens-per-second 246,259 | peak-gpu-memory 58.26 GB
2026-04-12 22:40:12 | INFO | step 00000015/00000015 | step-time 18.045 sec | cross-entropy-loss 2.5940 | learning-rate 1.000000e-06 | gradient-norm 937.0814 | tokens-per-second 232,441 | peak-gpu-memory 58.43 GB
```

File: pithtrain/dualpipe/execution.py [hunk 1]
```diff
@@ -28,6 +28,13 @@ class ExecutionCtx:
     bwd_event: torch.cuda.Event
     fwd_comm_work: Optional[torch.distributed.Work]
     bwd_comm_work: Optional[torch.distributed.Work]
+    fwd_comm_deferred_free: List[torch.Tensor]
+    """Tensors whose storage should be freed after the next fwd_comm_work.wait().
+
+    Callers append tensors here after launching async forward comms (e.g.
+    all-to-all in Stage 2 / Stage 4).  The subsequent stage that waits on
+    fwd_comm_work drains and frees this list automatically.
+    """
```
Comment (haok1402) [inline pithtrain/dualpipe/execution.py:45]: nit: other fields may benefit from similar documentations.

File: pithtrain/operators/all_to_all.py [hunk 2]
```diff
@@ -15,9 +15,10 @@
 # Pad all-to-all output allocations to multiples of this many rows.
 # Reduces CUDA memory fragmentation by collapsing varying allocation sizes
 # into the same caching-allocator bucket for reuse across micro-batches.
-_ALLTOALL_PAD_ALIGNMENT = 512
+_ALLTOALL_PAD_ALIGNMENT = 1024
 
 
+@torch.no_grad()
```
Comment (haok1402) [inline pithtrain/operators/all_to_all.py:21]: @MasterJH5574 Is this required? I believe before we pass the inputs for all2all, they are dispatched, which means the inputs will never require gradients.

https://docs.pytorch.org/docs/stable/generated/torch.Tensor.detach.html

File: tests/test_layer_partition.py
Comment (haok1402) [inline tests/test_layer_partition.py:1]: same comment ; whether layer partition is still needed

File: pithtrain/models/deepseek_v2_lite.py [hunk 3]
```diff
@@ -267,7 +262,7 @@ def compute(
         """
         _, _, h = hidden_states.shape
         hidden_states = hidden_states.view(-1, h)
-        logits = F.linear(hidden_states.type(torch.float32), self.weight.type(torch.float32), None)
```
Comment (haok1402) [inline pithtrain/models/deepseek_v2_lite.py:270]: @MasterJH5574 DeepSeek-V2-Lite, unlike Qwen3-30B-A3B, would run its router in FP32. We should probably revert back this change.

https://huggingface.co/deepseek-ai/DeepSeek-V2-Lite/blob/main/modeling_deepseek.py#L424-L426

**Review by haok1402 — APPROVED** [review]

File: pithtrain/dualpipe/layer_partition.py
Comment (MasterJH5574) [reply pithtrain/dualpipe/layer_partition.py:1]: I think it may be beneficial given the layer partition logic is expected to be the same?

File: pithtrain/operators/all_to_all.py (hunk 2 above)
Comment (MasterJH5574) [reply pithtrain/operators/all_to_all.py:21]: Yes that's true.  Just want to be explicit and make it crystal clear.

File: pithtrain/models/deepseek_v2_lite.py (hunk 3 above)
Comment (MasterJH5574) [reply pithtrain/models/deepseek_v2_lite.py:270]: Done. Thanks for pointing out.

File: pithtrain/dualpipe/execution.py (hunk 1 above)
Comment (MasterJH5574) [reply pithtrain/dualpipe/execution.py:45]: Added docstrings for other fields.

File: pithtrain/operators/cross_entropy.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        if inp.stride(-1) != 1 or inp.stride(-2) != n_cols:
+            inp = inp.contiguous()
+        if target.stride(-1) != 1:
+            target = target.contiguous()
+
+        n_non_ignore = (target != ignore_index).sum(dtype=torch.int64).view(1)
+        n_non_ignore.clamp_(min=1)
+
+        loss_1d = torch.zeros(n_rows, dtype=torch.float32, device=inp.device)
+        BLOCK_SIZE = min(65536 // 2, triton.next_power_of_2(n_cols))
```
[the remark this answers is not in the document]
Comment (MasterJH5574) [reply pithtrain/operators/cross_entropy.py:101]: This part is adapted from TransformerEngine https://github.com/NVIDIA/TransformerEngine/blob/2dd31bb849e83cce51c7d169db883862063d3a95/transformer_engine/pytorch/triton/cross_entropy.py#L23-L41
